torch/distributed/_shard/checkpoint/metadata.py

```diff
@@ -77,3 +78,25 @@ class TensorReadRequest:
     # offset and length w.r.t. to the storage identified by ``storage_key``
     offsets: Tuple[int, ...]
     lengths: Tuple[int, ...]
+
+
+@dataclass(frozen=True)
+class MetadataIndex:
+    """
+    This class represents a lookup key for items in a state dict or Metadata.
+    """
+    fqn: str
+    """Fully Qualified Name of the object"""
+
+    offset: Optional[torch.Size] = None
+    """If the object is a tensor, offset into the tensor we're looking for"""
+
+    index: Optional[int] = field(hash=False, compare=False, default=None)
```

Comment:
can we explain a bit how it could speed up look ups?

Reply:
Done. Added some comments on how to use it.
Basically it's caching the index into the list that holds the value you wanna fetch.